Show Jetpack banner in search results

ovitrif opened the pull request:
Show the Jetpack banner in search results to match iOS version.

To test:
Test Case 1 (Phone)
1. Launch WordPress App (in portrait orientation)
2. Go to Reader → Search
3. Enter a search query & submit (e.g. `Jetpack`)
4. **Expect** the Jetpack Powered banner to **hide** when the search is in progress. 
5. **Expect** the banner **show** when the search results appear.
6. Scroll down & **expect** the banner to **slide-out** of the view
7. Scroll back to the top and **expect** the banner to **slide-in** into the view
8. Rotate the phone to landscape
9. Tap the search input in the toolbar and change the search query (e.g. `Jetpack powered`)
10. **Expect** the banner to hide
11. Rotate the phone to portrait
12. **Expect** the banner to show 

Test Case 2 (Tablet)
- Repeat Test Case 1 but **expect** the banner to **show** when hidden in landscape orientation on a phone.
  Specifically that is when updating the search query.

<img width="314" src="https://user-images.githubusercontent.com/4588074/184150810-7515970c-a12c-4c94-b9e7-474bf2aac4e2.png">

## Regression Notes
1. Potential unintended areas of impact
   Site creation - Domain search **on a phone** → I've updated a function that wasn't working correctly to detect a phone in landscape orientation; making it not reusable in other places.
   That function was reused afterwards to hide the banner in reader search on a phone seen in landscape.

13. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing.

14. What automated tests I added (or what prevented me from doing so)
   N/a → UI code that doesn't warrant automated tests.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Show Jetpack banner in search results

ovitrif committed:
Update release notes for PR #17028

irfano reviewed (changes requested):
Good job @ovitrif! 
There's still a tiny difference between iOS. While searching, the banner is not visible on iOS. I think we can do the same for consistency.

The screenshot is from the iOS version of the latest [trunk](https://github.com/wordpress-mobile/WordPress-iOS/tree/3ff39e03d21a8387a58b9556c3a2492b994b17e8)

<img src="https://user-images.githubusercontent.com/2471769/184169310-43b1623d-5937-40c7-98b3-7e0e01a8bd65.png" width=240>

Since you add a note to `RELEASE-NOTES.txt`, you can check it on the PR submission checklist.

ovitrif commented:
> Good job @ovitrif! There's still a tiny difference between iOS. While searching, the banner is not visible on iOS. I think we can do the same for consistency.

Good point 👍 !
Thanks for reviewing and noticing this one @irfano 🙇. I'll look into a solution to align the behavior in this regard too.

> Since you add a note to `RELEASE-NOTES.txt`, you can check it on the PR submission checklist.

True story, thank you for the heads-up on this one too 🙇 !

ovitrif committed:
Refactor toggleJetpackBannerIfEnabled

ovitrif committed:
Show Jetpack banner when displaying search results, hide when loading

ovitrif committed:
Fix refactor from commit 38d8076d

ovitrif committed:
Convert View.OnClickListener() to lambda

ovitrif committed:
Set the navigation color for the Jetpack banner when toggling it

ovitrif committed:
Add slide in-out banner animation when scrolling search results

ovitrif commented:
@irfano I updated the implementation to hide the banner when loading the search results, and show it when the results appear on the screen.

I also applied the navigation bar coloring as it was missing in this fragment view, and implemented the animation on scroll 👍 .

There's also a bit of refactoring that emerged as I've understood this code a bit better ;)

irfano commented:
Thanks for the update, @ovitrif . I'll review the latest changes.

But I just noticed another small detail from the iOS version. On the iOS version, the banner is hidden on the reader search screen when the orientation is landscape. You can see it in this [PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/19122).

<img src="https://user-images.githubusercontent.com/2471769/184244272-34f6f1d4-c85e-4daf-b1f3-8f40f9811c7f.png" width=569>

ovitrif commented:
> But I just noticed another small detail from the iOS version. On the iOS version, the banner is hidden on the reader search screen when the orientation is landscape. You can see it in this [PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/19122).

Good tip, thank you for sharing this. I lost a bit of track of the Jetpack Focus work that's going on lately with the groundskeeping work that I've been into last 2 weeks and being more focused with analysis of the next project during the last couple of days.
It seems I still have some catching up to do 😅 

Thanks a lot for bringing points from the broader updated context in here @irfano 🙇 . I was totally unaware of this update; probably also because that PR is not merged yet.

**I'll see to it that I bring the hiding in landscape on Android too.**

While testing I noticed another issue with the current state of this PR, the bottom margin remains when the banner slides out of the view as we're scrolling down:

<img width="314" src="https://user-images.githubusercontent.com/4588074/184247662-d4988470-d95d-409f-9960-a73a72e6ac25.png">

Not sure if this issue was also happening in the emulator, but nonetheless it occurs on the physical device I'm testing with.

Edit: it's also reproducible on the emulator, as expected 😅 .

ovitrif committed:
Fix search results bottom padding remaining when jetpack banner slides out of view

ovitrif committed:
Hide Jetpack banner when searching in reader on a phone in landscape orientation

ovitrif commented:
@irfano I fixed both:
- The bottom padding of the list that remained when the banner was sliding out of view during scroll down 
   https://github.com/wordpress-mobile/WordPress-Android/pull/17028/commits/70c2b9e32e820f38f2db5a058a8613ff9999faaa
- Hidden the banner when updating the search query
  https://github.com/wordpress-mobile/WordPress-Android/pull/17028/commits/172e45254a683d765faf4e95a32e7cbb901047b0

irfano commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderSearchActivity.kt:
```diff
@@ -41,4 +47,23 @@ class ReaderSearchActivity : LocaleAwareActivity() {
         readerTracker.stop(MAIN_READER)
     }
+
+    override fun onScrollableViewInitialized(containerId: Int) {
+        val fragmentContainer = supportFragmentManager.findFragmentById(R.id.fragment_container)
+
+        if (fragmentContainer is ReaderPostListFragment) {
+            val fragmentView = fragmentContainer.view ?: return
+
+            fragmentView.post {
+                // post is used to create a minimal delay here. containerId changes just before
+                // onScrollableViewInitialized is called, and findViewById can't find the new id before the delay.
+                val jetpackBannerView = fragmentView.findViewById<View>(R.id.jetpack_banner)
+                val scrollableView = fragmentView.findViewById<View>(containerId) as RecyclerView
+
+                if (jetpackBrandingUtils.shouldShowJetpackBranding()) {
+                    jetpackBrandingUtils.initJetpackBannerAnimation(jetpackBannerView, scrollableView)
+                }
+            }
+        }
+    }
 }
```
I couldn't understand why we inform `ReaderSearchActivity` from `ReaderPostListFragment` about `onScrollableViewInitialized`. We don't do anything related to `ReaderSearchActivity` here, `jetpackBannerView` and `scrollableView` are both in `ReaderPostListFragment`. Can we move this to `ReaderPostListFragment`? 🤔

ovitrif replied:
I saw other activities that implement `ScrollableViewInitializedListener` using a similar approach, and I went with that although it's not needed.

By getting rid of `ScrollableViewInitializedListener` this should work directly from the fragment 👍

ovitrif replied:
Done in https://github.com/wordpress-mobile/WordPress-Android/pull/17028/commits/159f018cbc149575214d3727501dfc994e62150e

irfano reviewed (dismissed):
I tested different cases, and they all work well. Just some concerns:
- There is an extra redundant bottom margin of the search history list.
<img src="https://user-images.githubusercontent.com/2471769/184401063-cc6c3d67-6af4-48fb-8fd4-30b35b930c73.png" width=320>
- We can remove navigation bar color when search is in progres (So many cases 😅)
<img src="https://user-images.githubusercontent.com/2471769/184401250-ad39cce3-5666-4562-9224-775d1f4e583c.png" width=320>

ovitrif committed:
Init banner animation in onCreateView to get rid of ScrollableViewInitializedListener

ovitrif committed:
Remove bottom nav color when banner is hidden + fix redundant bottom margin in suggestions list

ovitrif commented:
> There is an extra redundant bottom margin of the search history list.
> - We can remove navigation bar color when search is in progres (So many cases 😅)

@irfano Fixed both issues in https://github.com/wordpress-mobile/WordPress-Android/pull/17028/commits/b5314d7dc7d7bf87054a7a99a15effeb9545cdfb

ravishanker reviewed (approved):
LGTM 👍

ovitrif commented:
Thank you @ravishanker for stepping in to review this PR in time ❤️ 🏅

AliSoftware commented:
@ovitrif @ravishanker While compiling the Release Notes for the 20.6 Code Freeze, I wanted to validate that the entry for this was supposed to be only for the WordPress release notes, but not for Jetpack, since iinm the Jetpack banner should only be shown on the WordPress app.

I just tested the Installable Build from this PR, and… **the Jetpack banner shows in the Jetpack app** once search results are shown (and it also doesn't disappear on scroll). We will probably want a betafix on the `release/20.6` branch for that.

<details><summary>🎦 Demo Video</summary>

https://user-images.githubusercontent.com/216089/185901123-d3d8a785-a183-4b89-ba34-bffc89605961.mp4

</details>

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java:
```diff
@@ -500,36 +504,49 @@ public void onCreate(Bundle savedInstanceState) {
-                    new JetpackPoweredBottomSheetFragment()
-                            .show(getChildFragmentManager(), JetpackPoweredBottomSheetFragment.TAG);
-                });
+        if (animateOnScroll) {
```
I believe this line bypasses the `mJetpackBrandingUtils.shouldShowJetpackBranding` check, which makes this banner visible in the Jetpack app, as described here https://github.com/wordpress-mobile/WordPress-Android/pull/17028#issuecomment-1222166253.

ovitrif replied:
Nicely spotted, that was indeed the problem. Thank you for fixing this in https://github.com/wordpress-mobile/WordPress-Android/pull/17062 🙇

mkevins replied:
This was fixed with: https://github.com/wordpress-mobile/WordPress-Android/pull/17062
